File: packages/sdk-js/src/integrations/alignment-state.ts
/**
 * Single-process buffer for the latest alignment verdict per
 * agent_id. Mirrors the Python `_alignment_state` module.
 *
 * The alignment helper records into this buffer; the auto-
 * instrumentation interceptor consumes it (single-use, TTL) when
 * building the next /check payload for that agent. DSL rules like
 * `{ alignment.drifted: true }` then fire on the same hop.
 */

export const TTL_MS = 30_000;

interface Entry {
  ts: number;
  verdict: Record<string, unknown>;
}

const state = new Map<string, Entry>();

export function record(agentId: string, verdict: Record<string, unknown>): void {
  if (!agentId || typeof verdict !== 'object' || verdict === null) return;
  state.set(agentId, { ts: Date.now(), verdict });
}

export function consume(agentId: string): Record<string, unknown> | null {
  const entry = state.get(agentId);
  if (!entry) return null;
  state.delete(agentId);
  if (Date.now() - entry.ts > TTL_MS) return null;
  return entry.verdict;
}

export function reset(): void {
  state.clear();
}

/**
 * Project a /alignment/check response down to the shape /check
 * accepts under the `alignment` field. The gateway zod schema
 * requires a numeric `score` (clamped to [0,1]); everything else
 * is optional. Unknown fields are dropped, signals are capped at
 * 5 entries × 40 chars to match the gateway bound.
 */
export function toCheckPayload(
  verdict: Record<string, unknown>,
): {
  score: number;
  drifted?: boolean;
  signals?: string[];
  reason?: string;
} | null {
  const scoreRaw = verdict.score;
  if (typeof scoreRaw !== 'number' || Number.isNaN(scoreRaw)) return null;
  const out: ReturnType<typeof toCheckPayload> = {
    score: Math.max(0, Math.min(1, scoreRaw)),
  };
  if (typeof verdict.drifted === 'boolean') out!.drifted = verdict.drifted;
  if (Array.isArray(verdict.signals)) {
    const signals = (verdict.signals as unknown[])
      .filter((s): s is string | number => typeof s === 'string' || typeof s === 'number')
      .map((s) => String(s).slice(0, 40))
      .slice(0, 5);
    if (signals.length) out!.signals = signals;
  }
  if (typeof verdict.reason === 'string') {
    out!.reason = verdict.reason.slice(0, 500);
  }
  return out;
}

File: packages/sdk-python/agentguard/integrations/_alignment_state.py
"""
Thread-safe storage for the latest alignment verdict per agent_id.

Why this exists: the LangChain `AlignmentCallback` runs on the
synchronous callback thread when a ReAct agent emits an action. The
SDK's auto-instrumentation interceptors (`auto.py`) run on whichever
thread or asyncio loop the user's Anthropic / OpenAI client lives on.
Those two paths may not share a Python execution context, so a plain
``ContextVar`` won't propagate.

Instead we keep a small module-level dict, keyed by ``agent_id``,
holding the most recent verdict and the time it was observed. The
auto-interceptor reads this when building the ``/check`` payload and
attaches the verdict if it's fresh (within ``TTL_SECONDS``). A stale
verdict from an earlier run is silently dropped.

Concurrency: the dict is guarded by a single ``threading.Lock``. All
operations are O(1) and the lock window is microseconds.
"""

from __future__ import annotations

import threading
import time
from typing import Any, Optional

TTL_SECONDS: float = 30.0
"""Maximum age (in seconds) for a verdict to still be considered
fresh. Past this, the auto-interceptor will treat it as missing and
fall back to running /check without an alignment signal."""

_lock = threading.Lock()
_state: dict[str, tuple[float, dict[str, Any]]] = {}


def record(agent_id: str, verdict: dict[str, Any]) -> None:
    """Store a verdict keyed by agent_id along with the wall-clock
    time it was recorded. Overwrites any previous entry."""
    if not agent_id or not isinstance(verdict, dict):
        return
    with _lock:
        _state[agent_id] = (time.time(), verdict)


def consume(agent_id: str) -> Optional[dict[str, Any]]:
    """Return the latest verdict for agent_id if it is younger than
    ``TTL_SECONDS``, then remove it from the store. Returns None if
    nothing is buffered or the entry is stale.

    "Consume once" semantics keep the /check payload tied to a single
    agent step — the next tool call gets a fresh alignment check, not
    a re-played stale verdict."""
    now = time.time()
    with _lock:
        entry = _state.pop(agent_id, None)
    if entry is None:
        return None
    ts, verdict = entry
    if now - ts > TTL_SECONDS:
        return None
    return verdict


def to_check_payload(verdict: dict[str, Any]) -> dict[str, Any]:
    """Pick the fields that the gateway's ``CheckRequestSchema``
    accepts under ``alignment``, dropping anything unexpected.

    The gateway zod schema (see check.ts) currently allows: score,
    drifted, signals (≤5, ≤40 chars each), reason (≤500 chars).
    """
    out: dict[str, Any] = {}
    score = verdict.get("score")
    if isinstance(score, (int, float)):
        # Clamp to [0,1] — the gateway will validate the range anyway.
        out["score"] = max(0.0, min(1.0, float(score)))
    drifted = verdict.get("drifted")
    if isinstance(drifted, bool):
        out["drifted"] = drifted
    signals = verdict.get("signals")
    if isinstance(signals, list):
        # Truncate each signal and cap list length to match gateway
        # validation. Keeps the SDK robust to model regressions.
        out["signals"] = [
            str(s)[:40] for s in signals if isinstance(s, (str, int, float))
        ][:5]
    reason = verdict.get("reason")
    if isinstance(reason, str):
        out["reason"] = reason[:500]
    return out


def reset() -> None:
    """Clear all stored verdicts. Intended for tests."""
    with _lock:
        _state.clear()
